File: python/day8.py
from utilities import read_lines
from typing import List


def check_instructions(instructions: List[str]) -> (int, bool):
    executed = [False] * len(instructions)
    curr_instruction_index = 0
    accumulator = 0
    while curr_instruction_index < len(instructions):
        if executed[curr_instruction_index]:
            # infinite loop happen
            return accumulator, False
        executed[curr_instruction_index] = True

        instruction = instructions[curr_instruction_index]
        if 'nop' in instruction:
            curr_instruction_index += 1
        elif 'jmp' in instruction:
            curr_instruction_index += int(instruction[4:])
        else:
            curr_instruction_index += 1
            accumulator += int(instruction[4:])

    # successfully terminated
    return accumulator, True


def try_terminate(instructions: List[str]) -> int:
    for i in range(len(instructions)):
        instruction = instructions[i]
        if 'nop' in instruction:
            instructions[i] = 'jmp' + instruction[3:]
            accumulator, terminated = check_instructions(instructions)
            if terminated:
                return accumulator
            instructions[i] = instruction
        elif 'jmp' in instruction:
            instructions[i] = 'nop' + instruction[3:]
            accumulator, terminated = check_instructions(instructions)
            if terminated:
                return accumulator
            instructions[i] = instruction

    return -1


def part_1(instructions: List[str]) -> int:
    accumulator, _ = check_instructions(instructions)
    return accumulator


def part_2(instructions: List[str]) -> int:
    return try_terminate(instructions)


if __name__ == '__main__':
    print('Day 8')
    instructions = read_lines('../inputs/day8.txt')
    print('\tPart 1: {}'.format(part_1(instructions)))
    print('\tPart 2: {}'.format(part_2(instructions)))

File: rust/src/day8.rs
use crate::utilities::read_lines;
use std::mem::replace;

fn execute_instructions(instructions: &Vec<String>) -> (i32, bool) {
    let mut executed = vec![false; instructions.len()];
    let mut current = 0;
    let mut accumulator = 0;
    while current < instructions.len() {
        if executed[current] {
            return (accumulator, false);
        }
        executed[current] = true;

        if instructions[current].contains("nop") {
            current += 1;
        } else if instructions[current].contains("jmp") {
            let value = instructions[current].as_str()[5..]
                .parse::<usize>()
                .unwrap();
            if instructions[current].contains('+') {
                current += value;
            } else {
                current -= value;
            }
        } else {
            accumulator += instructions[current].as_str()[4..].parse::<i32>().unwrap();
            current += 1;
        }
    }
    (accumulator, true)
}

fn change_instruction(instructions: &mut Vec<String>) -> i32 {
    let n = instructions.len();
    let instr_changes = [("jmp", "nop"), ("nop", "jmp")];
    for i in 0..n {
        for instr_change in &instr_changes {
            if instructions[i].contains(instr_change.0) {
                let mut new_instruction = instr_change.1.to_string();
                new_instruction.push_str(&instructions[i].as_str()[3..]);
                let old_instruction = replace(&mut instructions[i], new_instruction);
                let (accumulator, terminated) = execute_instructions(instructions);
                let _ = replace(&mut instructions[i], old_instruction);
                if terminated {
                    return accumulator;
                }
            }
        }
    }
    -1
}

fn part_1(instructions: &Vec<String>) {
    let (accumulator, _) = execute_instructions(instructions);
    println!("\tPart 1: {}", accumulator);
}

fn part_2(instructions: &mut Vec<String>) {
    println!("\tPart 2: {}", change_instruction(instructions));
}

pub fn main() {
    println!("Day 8");
    let mut instructions = read_lines("../inputs/day8.txt");
    part_1(&instructions);
    part_2(&mut instructions);
}
